pkg/operator/k8sutil/deployment.go

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2018 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package k8sutil
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/rook/rook/pkg/clusterd"
+	"k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func UpdateDeploymentAndWait(context *clusterd.Context, deployment *v1beta1.Deployment, namespace string) error {
+	original, err := context.Clientset.Extensions().Deployments(namespace).Get(deployment.Name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get deployment %s. %+v", deployment.Name, err)
+	}
+
+	logger.Infof("updating deployment %s", deployment.Name)
+	if _, err := context.Clientset.Extensions().Deployments(namespace).Update(deployment); err != nil {
+		return fmt.Errorf("failed to update deployment %s. %+v", deployment.Name, err)
+	}
+
+	// wait for the deployment to be restarted
+	sleepTime := 2 * time.Second
+	for i := 0; i < 30; i++ {
+		// check for the status of the deployment
+		d, err := context.Clientset.Extensions().Deployments(namespace).Get(deployment.Name, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to get deployment %s. %+v", deployment.Name, err)
+		}
+		if d.Status.ObservedGeneration != original.Status.ObservedGeneration && d.Status.UpdatedReplicas > 0 && d.Status.AvailableReplicas > 0 {
```

Comment:
updatedReplicas == readyReplicas?
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/types.go#L229

Reply:
@travisn I think `readyReplicas` should be used as this will make it easier when we introduce readiness and liveness probes.
The OSD won't count to `readyReplicas` unless the probe(s) are green*. Which is kind of good because having the operator "abort" on OSDs that are not ready is good so we don't kill a cluster.

If a Pod does not have any probes, it is "instant" ready.

Reply:
agreed, will change this so it will be more effective when we have the liveness probe